File: api/src/main/java/dev/ulloasp/mlsuite/admin/infrastructure/InfrastructureService.java
package dev.ulloasp.mlsuite.admin.infrastructure;

import org.springframework.stereotype.Service;

import com.fasterxml.jackson.databind.JsonNode;

@Service
public class InfrastructureService {

    private final OpsAgentClient opsAgentClient;

    public InfrastructureService(OpsAgentClient opsAgentClient) {
        this.opsAgentClient = opsAgentClient;
    }

    public JsonNode overview() {
        return opsAgentClient.overview();
    }

    public void action(String serviceName, ServiceActionRequest request) {
        opsAgentClient.action(serviceName, request.action());
    }

    public JsonNode logs(String serviceName, int tail) {
        return opsAgentClient.logs(serviceName, tail);
    }

    public TerminalSessionResponse createTerminalSession(TerminalSessionRequest request) {
        TerminalSessionResponse upstream = opsAgentClient.createTerminalSession(request);
        return new TerminalSessionResponse(upstream.sessionId(), "/api/admin/infrastructure/terminal/" + upstream.sessionId());
    }

    public void closeTerminalSession(String sessionId) {
        opsAgentClient.closeTerminalSession(sessionId);
    }
}

File: frontend/src/admin/infrastructure/api/infrastructureService.ts
import { appFetch } from "../../../app/api/appFetch";
import type {
	InfrastructureOverviewDto,
	ServiceLogsSnapshotDto,
	TerminalSessionDto,
} from "../types";

type ServiceAction = "START" | "STOP" | "RESTART";

export const getInfrastructureOverview = () =>
	appFetch<InfrastructureOverviewDto>("/api/admin/infrastructure/overview");

export const runServiceAction = (serviceName: string, action: ServiceAction) =>
	appFetch<void>(
		`/api/admin/infrastructure/services/${serviceName}/actions`,
		json("POST", { action }),
	);

export const getServiceLogsSnapshot = (serviceName: string, tail = 200) =>
	appFetch<ServiceLogsSnapshotDto>(
		`/api/admin/infrastructure/services/${serviceName}/logs?tail=${tail}`,
	);

export const createTerminalSession = (serviceName: string, cols: number, rows: number) =>
	appFetch<TerminalSessionDto>(
		"/api/admin/infrastructure/terminal/sessions",
		json("POST", { serviceName, cols, rows }),
	);

export const closeTerminalSession = (sessionId: string) =>
	appFetch<void>(`/api/admin/infrastructure/terminal/sessions/${sessionId}`, {
		method: "DELETE",
	});

function json(method: string, body: unknown): RequestInit {
	return {
		method,
		headers: { "Content-Type": "application/json" },
		body: JSON.stringify(body),
	};
}
